sdk/core/azure_core_amqp/src/value.rs

```diff
@@ -0,0 +1,735 @@
+    UByte(u8),
+    UShort(u16),
+    UInt(u32),
+    ULong(u64),
+    Byte(i8),
+    Short(i16),
+    Int(i32),
+    Long(i64),
+    Float(f32),
+    Double(f64),
+    Char(char),
+    TimeStamp(AmqpTimestamp),
+    Uuid(uuid::Uuid),
+    Binary(Vec<u8>),
+    String(String),
+    Symbol(AmqpSymbol),
+    List(AmqpList),
+    Map(AmqpOrderedMap<AmqpValue, AmqpValue>),
+    Array(Vec<AmqpValue>),
+    Described(Box<AmqpDescribed>),
+    Unknown,
+}
+
+impl AmqpValue {}
+
+impl<K, V> AmqpOrderedMap<K, V>
+where
+    K: PartialEq + Clone + Default,
+    V: Clone + Default,
+{
+    pub fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    pub fn insert(&mut self, key: impl Into<K>, value: impl Into<V>) {
+        self.inner.push((key.into(), value.into()));
+    }
+
+    pub fn get(&self, key: impl Into<K> + Clone) -> Option<&V> {
+        self.inner.iter().find_map(|(k, v)| {
+            if *k == key.clone().into() {
+                Some(v)
+            } else {
+                None
+            }
+        })
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        let index = self.inner.iter().position(|(k, _)| k == key)?;
+        Some(self.inner.remove(index).1)
+    }
+
+    pub fn contains_key(&self, key: impl Into<K> + Clone) -> bool {
+        self.inner.iter().any(|(k, _)| *k == key.clone().into())
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (K, V)> + '_ {
+        self.inner.iter().map(|(k, v)| (k.clone(), v.clone()))
+    }
+}
+
+impl<K, V> IntoIterator for AmqpOrderedMap<K, V>
+where
+    K: PartialEq + Default,
+    V: Default,
+{
+    type Item = (K, V);
+    type IntoIter = <Vec<(K, V)> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+macro_rules! conversions_for_amqp_types {
+    ($($t:ty, $field:ident),*) => {
+        $(
+            impl From<$t> for AmqpValue {
+                fn from(v: $t) -> Self {
+                    AmqpValue::$field(v)
+                }
+            }
+
+            impl From<AmqpValue> for $t {
+                fn from(v: AmqpValue) -> Self {
+                    match v {
+                        AmqpValue::$field(v) => v,
+                        _ => panic!("Expected a {}", stringify!($t)),
+                    }
+                }
+            }
+
+            impl PartialEq<$t> for AmqpValue {
+                fn eq(&self, other: &$t) -> bool {
+                    match self {
+                        AmqpValue::$field(v) => v == other,
+                        _ => false,
+                    }
+                }
+            }
+            impl PartialEq<AmqpValue> for $t {
+                fn eq(&self, other: &AmqpValue) -> bool {
+                    match other {
+                        AmqpValue::$field(v) => self == v,
+                        _ => false,
+                    }
+                }
+            }
+        )*
+    }
+}
+
+conversions_for_amqp_types!(
```

Comment:
Using function macros for ease: nice! But I'd consider something more like a tuple for input just to reinforce that there's a pairing here. This feels fragile. And, most likely, having the macro take a tuple argument matcher would actually make the impl easier as well (I haven't looked at it yet, mind you).

Reply:
For an example, see `create_enum` in `macros.rs` under `azure_core` in `main`.